### PR TITLE
L1 To L2 Transaction Queue

### PR DESCRIPTION
**Pull request by K-Ho**
## Description
Implementation of L1ToL2 Transaction Queue. I will eventually separate the RollupList into RollupQueue and RollupStack contracts to clean up this code, but will do so once I implement the rest of the queues/stacks and know what the best way to do that separation is.

## Questions
- Should we move the L1ToL2TransactionPasser into the `rollup-contracts` package?

### Fixes
- Fixes #YAS371

## Contributing Agreement

- [X] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      )
+      try {
+        // delete the single appended block
+        await l1ToL2TxQueue.dequeueBeforeInclusive(blockIndex)
+      } catch (err) {
+        // Success we threw an error!
+        return
+      }
+      throw new Error(
+        'Allowed non-canonicalTransactionChain account to dequeue block'
```
Comment (ben-chain) [inline packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts:148]: Same test util comment

File: packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      await l1ToL2TxQueue.connect(l1ToL2TransactionPasser).enqueueBlock(block) // Did not throw... success!
+    })
+    it('should not allow enqueue from other address', async () => {
+      const block = ['0x1234']
+      try {
+        await l1ToL2TxQueue.enqueueBlock(block)
+      } catch (err) {
+        // Success we threw an error!
+        return
+      }
```
Comment (ben-chain) [inline packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts:88]: Note we do have a `TestUtils.assertThrowsAsync` which replicates this functionality :)

Comment (willmeister) [reply packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts:88]: Would also check the error that is thrown to make sure it is doing the right thing vs failing for some other reason when doing the wrong thing.

File: packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+/* External Imports */
+import { getLogger } from '@eth-optimism/core-utils'
+import { createMockProvider, deployContract, getWallets } from 'ethereum-waffle'
+
+/* Internal Imports */
+import { DefaultRollupBlock } from './RLhelper'
+
+/* Logging */
+const log = getLogger('l1-to-l2-tx-queue')
```
Comment (willmeister) [inline packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts:11]:
```suggestion
const log = getLogger('l1-to-l2-tx-queue', true)
```

Comment (ben-chain) [reply packages/rollup-contracts/test/rollup-list/L1ToL2TransactionQueue.spec.ts:11]: Oh yes good point, come to think of it I think there is an `assertRevertsAsync` which accepts an EVM revert message to check!

**Pull request merged**